crates/router/src/db/address.rs

```diff
@@ -105,85 +107,341 @@ impl AddressInterface for Store {
+                    payment_id,
+                    address_id,
+                )
+                .await
+                .map_err(Into::into)
+                .into_report()
+            };
+            let address = match storage_scheme {
+                MerchantStorageScheme::PostgresOnly => database_call().await,
+                MerchantStorageScheme::RedisKv => {
+                    let key = format!("{}_{}", merchant_id, payment_id);
+                    let field = format!("add_{}", address_id);
+                    db_utils::try_redis_get_else_try_database_get(
+                        self.get_redis_conn()
+                            .change_context(errors::StorageError::DatabaseConnectionError)?
+                            .get_hash_field_and_deserialize(&key, &field, "Address"),
+                        database_call,
+                    )
                     .await
-                    .change_context(errors::StorageError::DecryptionError)
-            })
-            .await
-    }
+                }
+            }?;
+            address
+                .convert(key_store.key.get_inner())
+                .await
+                .change_context(errors::StorageError::DecryptionError)
+        }
 
-    async fn update_address_by_merchant_id_customer_id(
-        &self,
-        customer_id: &str,
-        merchant_id: &str,
-        address: storage::AddressUpdate,
-        key_store: &domain::MerchantKeyStore,
-    ) -> CustomResult<Vec<domain::Address>, errors::StorageError> {
-        let conn = connection::pg_connection_write(self).await?;
-        storage::Address::update_by_merchant_id_customer_id(
-            &conn,
-            customer_id,
-            merchant_id,
-            address.into(),
-        )
-        .await
-        .map_err(Into::into)
-        .into_report()
-        .async_and_then(|addresses| async {
-            let mut output = Vec::with_capacity(addresses.len());
-            for address in addresses.into_iter() {
-                output.push(
+        #[instrument(skip_all)]
+        async fn update_address(
+            &self,
+            address_id: String,
+            address: storage_types::AddressUpdate,
+            key_store: &domain::MerchantKeyStore,
+        ) -> CustomResult<domain::Address, errors::StorageError> {
+            let conn = connection::pg_connection_write(self).await?;
+            storage_types::Address::update_by_address_id(&conn, address_id, address.into())
+                .await
+                .map_err(Into::into)
+                .into_report()
+                .async_and_then(|address| async {
                     address
                         .convert(key_store.key.get_inner())
                         .await
-                        .change_context(errors::StorageError::DecryptionError)?,
-                )
+                        .change_context(errors::StorageError::DecryptionError)
+                })
+                .await
```

Comment:
Shouldn't you redact the previous value in redis while updating.

Reply:
Currently we have KV implementation only for address linked to payment flows! In payments, we only insert and read address but we do not update address! Customer related address flows has insert and update where we did not implement KV